src/compiler/checker.ts

```diff
@@ -26021,7 +26021,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function eachTypeContainedIn(source: Type, types: Type[]) {
-        return source.flags & TypeFlags.Union ? !forEach((source as UnionType).types, t => !contains(types, t)) : contains(types, source);
+        const equalityComparer = (sourceType: Type, targetType: Type) =>
+            sourceType === targetType || !!(sourceType.flags & TypeFlags.Literal && targetType.flags & TypeFlags.Literal && (sourceType as LiteralType).value === (targetType as LiteralType).value);

```

Comment:
I'm not a total expert, but my gut tells that this isn't the right place to do this, given this is presumably a hot function that previously did something very simple (check if one type is in another by reference only), but now also encodes some niche literal equality behavior.

I feel like I would need to debug and see who's calling this to understand why this is the function to change and not something else...

Reply:
Ah, this isn't the function I was thinking of; it's named generically but is only used once in switch case checking. Maybe it hasn't always been that way?

Doing some sleuthing, and this was introduced in #9163 and all references but one were removed in #9407 (https://github.com/microsoft/TypeScript/commit/280b27e2630e1fae6868492613431819ab40f18c), then never touched again. So, it has been this way for a very long time, it seems.

Not sure what that means for this, but I still somewhat feel like this isn't quite the right place. Maybe the caller's `mapType` would be better off mapping things differently? Unsure.

@ahejlsberg probably knows better or maybe @weswigham

Reply:
The thing I wonder is that the linked PRs showed code that used this function and then switched away to another method of checking things, which makes me feel like there's some other more-robust-and-consistent way that this would be done if it were reimplemented today, e.g. without this simple exact equality check.

Reply:
@weswigham It seems that `eachTypeContainedIn` is only used in `computeExhaustiveSwitchStatement`. Perhaps we can inline it there?

Reply:
Should the `LiteralType` be handled differently for the enums? I'm just trying to figure out in which direction to investigate this issue...

Reply:
Handled where?

I think the thing I was hoping was that this function could be dropped entirely in place of some relation + `getUnionType(switchTypes)` calls in `computeExhaustiveSwitchStatement`, but I'm having trouble getting it to work by flailing around.

Mainly, I'm just trying to figure out if there's existing code which can already do this narrowing e.g. used by plain ol if statements or something.